script/entity_linker.py

```diff
@@ -0,0 +1,953 @@
+        task="token-classification",
+        model=cer_model,
+        tokenizer=cer_tokenizer,
+        aggregation_strategy="first",
+        device="cpu",
+    )
+    return cer_pipeline
+
+
+def main():
+    # columns are note_id, text
+    logger.info("Reading in notes data.")
+    notes = pd.read_csv(NOTES_PATH)
+    logger.info(f"Found {notes.shape[0]} notes.")
+    spans = []
+
+    # Load model components
+    logger.info("Loading CER pipeline.")
+    cer_pipeline = load_cer_pipeline()
+    cer_tokenizer = cer_pipeline.tokenizer
+
+    logger.info("Loading Linker")
+    with open(LINKER_PATH, "rb") as f:
+        linker = pickle.load(f)
+
+    # Process one note at a time...
+    logger.info("Processing notes.")
+    for row in notes.itertuples():
+        # Tokenize the entire discharge note
+        tokenized = cer_tokenizer(
+            row.text,
+            return_offsets_mapping=False,
+            add_special_tokens=False,
+            truncation=False,
+        )
+
+        global_offset = 0
+        global_offset_mapping = []
+
+        # Adjust the token offsets so that they match the original document
+        for input_id in tokenized["input_ids"]:
+            token = cer_tokenizer.decode(input_id)
+            pos = row.text[global_offset:].find(token)
+            start = global_offset + pos
+            end = global_offset + pos + len(token)
+            global_offset = end
+            global_offset_mapping.append((start, end))
+
+        chunk_start_idx = 0
+        chunk_end_idx = 0
+
+        # Process the document in chunks of 512 tokens chunk at a time
+        for offset_chunk in chunked(global_offset_mapping, MAX_SEQ_LEN - 2):
+            chunk_start_idx = chunk_end_idx
+            chunk_end_idx = offset_chunk[-1][1]
+            chunk_text = row.text[chunk_start_idx:chunk_end_idx]
+
+            # ...one matched clinical entity at a time
+            # Iterate through the detected entities and link them
+            for entity in cer_pipeline(chunk_text):
+                example = pd.Series(
+                    {
+                        # +1 to account for the [CLS] token
+                        "start": entity["start"] + chunk_start_idx + 1,
+                        "end": entity["end"] + chunk_start_idx,
+                        "text": row.text,
+                    }
+                )
+                sctid = linker.link(example)
+                if sctid:
+                    spans.append(
+                        {
+                            "note_id": row.Index,
+                            "start": example["start"],
+                            "end": example["end"],
+                            "concept_id": sctid,
+                        }
+                    )
+
+    logger.info(f"Generated {len(spans)} annotated spans.")
+    spans_df = pd.DataFrame(spans)
+    spans_df.to_csv(SUBMISSION_PATH, index=False)
+    logger.info("Finished.")
+
+
+if __name__ == "__main__":
+    main()
+
+
```

Comment:
I've got the blurb, but the screenshot is hard since I need a CUDA machine to run inference. Would a screenshot of the benchmark screen in the blog post suffice?

Reply:
It should just be a screenshot from the competition workflow, something in the vein of the below examples

airports example
<img width="952" alt="image" src="https://github.com/BayesRulez/snomed_el_baseline_model/assets/22667367/1e5b54e5-fc8e-428a-88d0-377f8ea89671">


visiomel example
<img width="964" alt="image" src="https://github.com/BayesRulez/snomed_el_baseline_model/assets/22667367/c614c2c2-ef39-4fb1-b731-eea103169391">

Reply:
I can add this to the blog post itself on dd-labs, i don't think it needs to be present here

Reply:
agreed